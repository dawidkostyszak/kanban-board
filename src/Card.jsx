import React from 'react';
import PropTypes from 'prop-types';

import connect from './Store.jsx';

export class StatelessCard extends React.Component {
    saveCardName = (input) => {
        this.props.updateCardName(input.target.value);
    }

    addCard = () => {
        const { card: { name } } = this.props;
        if (name.trim().length) {
            this.props.saveCard();
        }
    }

    removeCard = () => {
        this.props.removeCard();
    }

    renderCard () {
        const { card: { name } } = this.props;
        return (
            <li className="c-card">
                <span className="c-card__title">{name}</span>
            </li>
        );
    }

    renderNewCard () {
        const { card: { name } } = this.props;
        return (
            <div className="c-new-card">
                <textarea
                    className="c-input c-new-card__input"
                    onChange={this.saveCardName}
                    placeholder="Add a card..."
                    value={name}
                />
                <button
                    className="c-button c-button--save"
                    onClick={this.addCard}
                >
                    Add
                </button>
                <button
                    className="c-button c-button--cancel"
                    onClick={this.removeCard}
                >
                    X
                </button>
            </div>
        );
    }

    render () {
        const { card: { isNew } } = this.props;
        if (isNew) {
            return this.renderNewCard();
        }
        return this.renderCard();
    }
}

StatelessCard.propTypes = {
    saveCard: PropTypes.func,
    removeCard: PropTypes.func,
    updateCardName: PropTypes.func,
    card: PropTypes.shape({
        name: PropTypes.string,
        isNew: PropTypes.bool,
    }),
};

const mapDispatchToProps = (dispatch, state, props) => ({
    updateCardName: (name) => {
        const lists = state.lists.map(
            (list) => {
                if (list.id === props.listId) {
                    return {
                        ...list,
                        cards: list.cards.map(
                            (card) => {
                                if (card.id === props.card.id) {
                                    return {
                                        ...card,
                                        name,
                                    };
                                }
                                return card;
                            },
                        ),
                    };
                }
                return list;
            },
        );

        dispatch({
            ...state,
            lists,
        });
    },
    saveCard: () => {
        const lists = state.lists.map(
            (list) => {
                if (list.id === props.listId) {
                    return {
                        ...list,
                        cards: list.cards.map(
                            (card) => {
                                if (card.id === props.card.id) {
                                    return {
                                        ...card,
                                        isNew: false,
                                    };
                                }
                                return card;
                            },
                        ),
                        newCardAdded: false,
                    };
                }
                return list;
            },
        );

        dispatch({
            ...state,
            lists,
        });
    },
    removeCard: () => {
        const lists = state.lists.map(
            (list) => {
                if (list.id === props.listId) {
                    return {
                        ...list,
                        cards: list.cards.filter(
                            (card) => card.id !== props.card.id,
                        ),
                        newCardAdded: false,
                    };
                }
                return list;
            },
        );

        dispatch({
            ...state,
            lists,
        });
    },
});

export default connect(StatelessCard, { mapDispatchToProps });
