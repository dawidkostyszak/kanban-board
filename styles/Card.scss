.c-card {
  background-color: #ffffff;
  list-style: none;
  border-radius: 5px;
  margin: 0.25rem 0.5rem;
  padding: 0.5rem;

  &__title {
    color: #4d4d4d;
    font-size: 1.5rem;
  }
}