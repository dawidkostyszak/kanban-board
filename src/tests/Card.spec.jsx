import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { StatelessCard } from '../Card.jsx';

describe('Card component', () => {
    let props;

    beforeEach(() => {
        props = {
            saveCard: jest.fn(),
            removeCard: jest.fn(),
            updateCardName: jest.fn(),
            card: {
                name: 'Test',
                isNew: false,
            },
        };
    });

    it('should render component correctly', () => {
        const wrapper = shallow(<StatelessCard {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render input for card name', () => {
        props.card = {
            name: '',
            isNew: true,
        };
        const wrapper = shallow(<StatelessCard {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should call updateCardName on input change', () => {
        props.card = {
            name: '',
            isNew: true,
        };
        const wrapper = shallow(<StatelessCard {...props} />);
        wrapper.find('.c-new-card__input').simulate(
            'change',
            { target: { value: 'New value' } },
        );
        expect(props.updateCardName).toHaveBeenCalled;
    });

    it('should not call addCard on save button click when name is empty', () => {
        props.card = {
            name: '',
            isNew: true,
        };
        const wrapper = shallow(<StatelessCard {...props} />);
        wrapper.find('.c-button--save').simulate('click');
        expect(props.addCard).not.toHaveBeenCalled;
    });

    it('should call addCard on save button click when value is not empty', () => {
        props.card = {
            name: 'Test',
            isNew: true,
        };
        const wrapper = shallow(<StatelessCard {...props} />);
        wrapper.find('.c-button--save').simulate('click');
        expect(props.addCard).toHaveBeenCalled;
    });

    it('should call removeCard on cancel button click', () => {
        props.card = {
            name: 'Test',
            isNew: true,
        };
        const wrapper = shallow(<StatelessCard {...props} />);
        wrapper.find('.c-button--cancel').simulate('click');
        expect(props.removeCard).toHaveBeenCalled;
    });
});
