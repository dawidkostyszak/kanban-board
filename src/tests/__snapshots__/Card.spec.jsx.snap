// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render component correctly 1`] = `
<li
  className="c-card"
>
  <span
    className="c-card__title"
  >
    Test
  </span>
</li>
`;

exports[`Card component should render input for card name 1`] = `
<div
  className="c-new-card"
>
  <textarea
    className="c-input c-new-card__input"
    onChange={[Function]}
    placeholder="Add a card..."
    value=""
  />
  <button
    className="c-button c-button--save"
    onClick={[Function]}
  >
    Add
  </button>
  <button
    className="c-button c-button--cancel"
    onClick={[Function]}
  >
    X
  </button>
</div>
`;
